File: android/MedicalFactChecker/app/src/main/java/com/bmlibrarian/factchecker/domain/workflow/ParallelScoringService.kt
/*
 * BMLibrarian Lite - Biomedical Literature Research Tool
 * Copyright (C) 2024-2025 Dr Horst Herb
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package com.bmlibrarian.factchecker.domain.workflow

import com.bmlibrarian.factchecker.data.remote.llm.LLMService
import com.bmlibrarian.factchecker.domain.model.LLMProvider
import com.bmlibrarian.factchecker.domain.model.ModelInfo
import com.bmlibrarian.factchecker.util.Constants
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.sync.Semaphore
import kotlinx.coroutines.sync.withPermit
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.math.pow
import kotlin.random.Random

/**
 * Service for parallel document scoring using Kotlin Coroutines.
 *
 * Uses a Semaphore to limit concurrent scoring requests while processing
 * documents in parallel. Results are reported via progress callbacks as
 * they complete.
 *
 * Thread Safety:
 * - Input uses ScoringInput data classes (not Room entities)
 * - Output uses ScoringResult sealed classes
 * - The calling workflow applies results to DocumentEntities on the main thread
 *
 * Mirrors iOS ParallelScoringService for cross-platform consistency.
 *
 * Example usage:
 * ```kotlin
 * val service = ParallelScoringService(llmService)
 *
 * val inputs = documents.map { ScoringInput.fromDocument(it) }
 *
 * val results = service.scoreDocuments(
 *     documents = inputs,
 *     claim = "Vitamin D reduces COVID-19 severity",
 *     provider = LLMProvider.OPENAI,
 *     apiKey = "...",
 *     model = "gpt-4o",
 *     maxConcurrent = 3
 * ) { documentId, completed, total ->
 *     println("Scored $completed/$total")
 * }
 * ```
 */
@Singleton
class ParallelScoringService @Inject constructor(
    private val llmService: LLMService
) {
    companion object {
        /** Maximum parse retries per document before giving up. */
        private const val MAX_PARSE_RETRIES = 3

        /** Base delay for parse retry backoff (milliseconds). */
        private const val PARSE_RETRY_BASE_DELAY_MS = 500L

        /** Default max concurrent requests for cloud providers. */
        const val DEFAULT_CLOUD_CONCURRENCY = 3

        /** Max concurrent requests for local providers (Ollama). */
        const val LOCAL_CONCURRENCY = 1
    }

    /**
     * Score multiple documents in parallel.
     *
     * Uses a semaphore-based approach to maintain exactly `maxConcurrent`
     * requests in flight at any time. Results are returned as they complete.
     *
     * @param documents Documents to score (as ScoringInput objects).
     * @param claim The medical claim to evaluate documents against.
     * @param provider The LLM provider configuration.
     * @param apiKey API key for authentication.
     * @param model Model ID to use.
     * @param maxConcurrent Maximum concurrent scoring requests.
     * @param onProgress Callback for progress updates (documentId, completed, total).
     * @return List of scoring results in completion order.
     */
    suspend fun scoreDocuments(
        documents: List<ScoringInput>,
        claim: String,
        provider: LLMProvider,
        apiKey: String,
        model: String,
        maxConcurrent: Int = detectConcurrency(provider),
        onProgress: suspend (String, Int, Int) -> Unit = { _, _, _ -> }
    ): List<ScoringResult> = coroutineScope {
        if (documents.isEmpty()) return@coroutineScope emptyList()

        val semaphore = Semaphore(maxConcurrent.coerceAtLeast(1))
        val total = documents.size
        var completed = 0

        val modelInfo = provider.getModel(model)

        // Launch all scoring tasks with semaphore-limited concurrency
        val deferredResults = documents.map { input ->
            async {
                semaphore.withPermit {
                    if (!isActive) {
                        // Return a cancellation failure if the coroutine was cancelled
                        return@withPermit ScoringResult.failure(
                            documentId = input.documentId,
                            pmid = input.pmid,
                            error = kotlinx.coroutines.CancellationException("Scoring cancelled"),
                            isRetryable = false
                        )
                    }

                    val result = scoreDocument(
                        input = input,
                        claim = claim,
                        provider = provider,
                        apiKey = apiKey,
                        model = model,
                        modelInfo = modelInfo
                    )

                    // Update progress
                    synchronized(this@coroutineScope) {
                        completed++
                    }
                    onProgress(input.documentId, completed, total)

                    result
                }
            }
        }

        // Await all results
        deferredResults.awaitAll()
    }

    /**
     * Score documents with checkpoint support.
     *
     * Skips documents that have already been scored (via checkpointedIds).
     * Results from checkpoints can be provided to include in aggregation.
     *
     * @param documents All documents to potentially score.
     * @param claim The medical claim to evaluate documents against.
     * @param provider The LLM provider configuration.
     * @param apiKey API key for authentication.
     * @param model Model ID to use.
     * @param checkpointedIds Set of document IDs that have been checkpointed.
     * @param maxConcurrent Maximum concurrent scoring requests.
     * @param onProgress Callback for progress updates.
     * @param onResult Callback for each result (for checkpointing).
     * @return List of scoring results for newly scored documents.
     */
    suspend fun scoreDocumentsWithCheckpoints(
        documents: List<ScoringInput>,
        claim: String,
        provider: LLMProvider,
        apiKey: String,
        model: String,
        checkpointedIds: Set<String>,
        maxConcurrent: Int = detectConcurrency(provider),
        onProgress: suspend (String, Int, Int) -> Unit = { _, _, _ -> },
        onResult: suspend (ScoringResult) -> Unit = { }
    ): List<ScoringResult> = coroutineScope {
        // Filter out already checkpointed documents
        val docsToScore = documents.filter { it.documentId !in checkpointedIds }

        if (docsToScore.isEmpty()) return@coroutineScope emptyList()

        val semaphore = Semaphore(maxConcurrent.coerceAtLeast(1))
        val totalRemaining = docsToScore.size
        val totalWithCheckpointed = documents.size
        val checkpointedCount = checkpointedIds.size
        var completed = 0

        val modelInfo = provider.getModel(model)

        val deferredResults = docsToScore.map { input ->
            async {
                semaphore.withPermit {
                    if (!isActive) {
                        return@withPermit ScoringResult.failure(
                            documentId = input.documentId,
                            pmid = input.pmid,
                            error = kotlinx.coroutines.CancellationException("Scoring cancelled"),
                            isRetryable = false
                        )
                    }

                    val result = scoreDocument(
                        input = input,
                        claim = claim,
                        provider = provider,
                        apiKey = apiKey,
                        model = model,
                        modelInfo = modelInfo
                    )

                    // Callback for checkpointing
                    onResult(result)

                    // Update progress (including checkpointed count)
                    synchronized(this@coroutineScope) {
                        completed++
                    }
                    onProgress(
                        input.documentId,
                        checkpointedCount + completed,
                        totalWithCheckpointed
                    )

                    result
                }
            }
        }

        deferredResults.awaitAll()
    }

    /**
     * Score a single document with retry logic for parse failures.
     *
     * @param input The document scoring input.
     * @param claim The claim to evaluate against.
     * @param provider The LLM provider.
     * @param apiKey API key for authentication.
     * @param model Model ID to use.
     * @param modelInfo Model pricing info for cost calculation.
     * @return Scoring result with score/rationale or error.
     */
    private suspend fun scoreDocument(
        input: ScoringInput,
        claim: String,
        provider: LLMProvider,
        apiKey: String,
        model: String,
        modelInfo: ModelInfo?
    ): ScoringResult {
        var lastParseError = ""
        var accumulatedInputTokens = 0
        var accumulatedOutputTokens = 0
        var accumulatedCost = 0.0

        // Retry loop for parse failures
        for (attempt in 0 until MAX_PARSE_RETRIES) {
            val result = llmService.scoreDocument(
                provider = provider,
                apiKey = apiKey,
                model = model,
                claim = claim,
                title = input.title,
                abstractText = input.abstract.ifEmpty { null }
            )

            // Estimate tokens and cost
            val estimatedInput = estimateTokens(claim + input.title + input.abstract)
            val estimatedOutput = Constants.LLM_SCORING_MAX_TOKENS / Constants.OUTPUT_TOKEN_ESTIMATE_DIVISOR
            val cost = modelInfo?.calculateCost(estimatedInput, estimatedOutput) ?: 0.0

            accumulatedInputTokens += estimatedInput
            accumulatedOutputTokens += estimatedOutput
            accumulatedCost += cost

            result.fold(
                onSuccess = { (score, rationale) ->
                    return ScoringResult.success(
                        documentId = input.documentId,
                        pmid = input.pmid,
                        score = score,
                        rationale = rationale,
                        inputTokens = accumulatedInputTokens,
                        outputTokens = accumulatedOutputTokens,
                        costUsd = accumulatedCost
                    )
                },
                onFailure = { error ->
                    // Check if this is a parse error (retryable) or API error (not retryable here)
                    val errorMessage = error.message ?: "Unknown error"

                    if (isParseError(error)) {
                        lastParseError = errorMessage

                        // Retry with exponential backoff
                        if (attempt < MAX_PARSE_RETRIES - 1) {
                            val delayMs = PARSE_RETRY_BASE_DELAY_MS * 2.0.pow(attempt.toDouble()).toLong()
                            val jitter = (delayMs * Random.nextDouble(-0.25, 0.25)).toLong()
                            delay(delayMs + jitter)
                        }
                    } else {
                        // API/network error - return immediately
                        return ScoringResult.failure(
                            documentId = input.documentId,
                            pmid = input.pmid,
                            error = error,
                            isRetryable = isRetryableError(error),
                            inputTokens = accumulatedInputTokens,
                            outputTokens = accumulatedOutputTokens,
                            costUsd = accumulatedCost
                        )
                    }
                }
            )
        }

        // All parse retries exhausted
        return ScoringResult.parseFailure(
            documentId = input.documentId,
            pmid = input.pmid,
            rawResponse = "",
            parseError = lastParseError,
            inputTokens = accumulatedInputTokens,
            outputTokens = accumulatedOutputTokens,
            costUsd = accumulatedCost
        )
    }

    /**
     * Detect appropriate concurrency level based on provider.
     *
     * Local providers (Ollama) use single concurrency, while cloud
     * providers can handle multiple concurrent requests.
     *
     * @param provider The LLM provider.
     * @return Recommended concurrency level.
     */
    fun detectConcurrency(provider: LLMProvider): Int {
        return if (provider.requiresApiKey) {
            DEFAULT_CLOUD_CONCURRENCY
        } else {
            LOCAL_CONCURRENCY
        }
    }

    /**
     * Check if an error is a parse error (response parsing failed).
     */
    private fun isParseError(error: Throwable): Boolean {
        val message = error.message?.lowercase() ?: return false
        return message.contains("parse") ||
                message.contains("json") ||
                message.contains("format") ||
                message.contains("invalid score")
    }

    /**
     * Check if an error is retryable (transient network issues).
     */
    private fun isRetryableError(error: Throwable): Boolean {
        val message = error.message?.lowercase() ?: return true
        return message.contains("timeout") ||
                message.contains("connection") ||
                message.contains("network") ||
                message.contains("rate limit") ||
                message.contains("429") ||
                message.contains("503") ||
                message.contains("502")
    }

    /**
     * Estimate token count for text.
     *
     * Uses a rough approximation based on average characters per token.
     *
     * @param text The text to estimate tokens for.
     * @return Estimated token count (minimum 1).
     */
    private fun estimateTokens(text: String): Int {
        return (text.length / Constants.TOKEN_ESTIMATE_CHARS_PER_TOKEN).coerceAtLeast(1)
    }
}

File: ios/MedicalFactChecker/Sources/Services/ParallelScoringService.swift
// BMLibrarian Lite - Biomedical Literature Research Tool
// Copyright (C) 2024-2026 Dr Horst Herb
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

import BioMedLit
import Foundation

/// Input data for scoring a single document.
///
/// Contains only the fields needed for the scoring prompt, avoiding
/// direct reference to SwiftData model objects for thread safety.
struct ScoringInput: Sendable {
    /// Unique identifier for matching back to the Document.
    let pmid: String

    /// Document title for the scoring prompt.
    let title: String

    /// Document abstract for the scoring prompt.
    let abstract: String

    /// Formatted author string for the scoring prompt.
    let authors: String

    /// Publication year (0 if unknown).
    let year: Int

    /// Journal name (or "Unknown" if not available).
    let journal: String

    /// Create scoring input from document fields.
    ///
    /// - Parameters:
    ///   - pmid: PubMed identifier.
    ///   - title: Document title.
    ///   - abstract: Document abstract.
    ///   - authors: Formatted author string (e.g., "Smith et al.").
    ///   - year: Publication year.
    ///   - journal: Journal name.
    init(
        pmid: String,
        title: String,
        abstract: String,
        authors: String,
        year: Int,
        journal: String
    ) {
        self.pmid = pmid
        self.title = title
        self.abstract = abstract
        self.authors = authors
        self.year = year
        self.journal = journal
    }
}

/// Result of scoring a single document.
///
/// Lightweight struct containing only essential data for updating the
/// Document model after scoring completes. Fully `Sendable` for safe
/// use across actor boundaries.
struct ScoringResult: Sendable {
    /// Document PMID for matching back to the source Document.
    let pmid: String

    /// Relevance score (1-5), nil if scoring failed.
    let score: Int?

    /// Explanation/rationale for the score.
    let rationale: String?

    /// Error message if scoring failed (stored as String for Sendable conformance).
    let errorMessage: String?

    /// Token usage for this scoring request.
    let usage: LLMUsage?

    /// Whether scoring failed due to an error.
    var isError: Bool { errorMessage != nil }

    /// Whether scoring succeeded with a valid score.
    var isSuccess: Bool { score != nil && errorMessage == nil }

    /// Create a successful scoring result.
    static func success(
        pmid: String,
        score: Int,
        rationale: String,
        usage: LLMUsage?
    ) -> ScoringResult {
        ScoringResult(
            pmid: pmid,
            score: score,
            rationale: rationale,
            errorMessage: nil,
            usage: usage
        )
    }

    /// Create a failed scoring result.
    static func failure(
        pmid: String,
        error: Error,
        usage: LLMUsage?
    ) -> ScoringResult {
        ScoringResult(
            pmid: pmid,
            score: nil,
            rationale: nil,
            errorMessage: error.localizedDescription,
            usage: usage
        )
    }

    /// Create a parse failure result (LLM responded but couldn't parse score).
    static func parseFailure(
        pmid: String,
        message: String,
        usage: LLMUsage?
    ) -> ScoringResult {
        ScoringResult(
            pmid: pmid,
            score: nil,
            rationale: message,
            errorMessage: "Failed to parse score: \(message)",
            usage: usage
        )
    }
}

/// Errors specific to document scoring.
enum ScoringError: LocalizedError, Sendable {
    case parseFailed(String)
    case allRetriesFailed(String)

    var errorDescription: String? {
        switch self {
        case .parseFailed(let message):
            return "Failed to parse score: \(message)"
        case .allRetriesFailed(let message):
            return "All scoring retries failed: \(message)"
        }
    }
}

/// Service for parallel document scoring using Swift Concurrency.
///
/// Uses `TaskGroup` to process multiple documents concurrently while
/// maintaining a configurable maximum concurrency level. Results are
/// reported via progress callbacks as they complete.
///
/// ## Thread Safety
///
/// This actor is designed for safe concurrent use:
/// - Input uses `ScoringInput` structs (not SwiftData models)
/// - Output uses `ScoringResult` structs (fully Sendable)
/// - The calling workflow applies results to Documents on the main actor
///
/// ## Example Usage
///
/// ```swift
/// let service = ParallelScoringService(llmService: llmService, maxConcurrent: 3)
///
/// let inputs = documents.map { doc in
///     ScoringInput(
///         pmid: doc.pmid,
///         title: doc.title,
///         abstract: doc.abstract,
///         authors: doc.formattedAuthors,
///         year: doc.year ?? 0,
///         journal: doc.journal ?? "Unknown"
///     )
/// }
///
/// let results = await service.scoreDocuments(
///     inputs,
///     claim: "Vitamin D reduces COVID-19 severity",
///     onProgress: { pmid, completed, total in
///         print("Scored \(completed)/\(total)")
///     }
/// )
/// ```
actor ParallelScoringService {
    // MARK: - Constants

    /// Maximum parse retries per document before giving up.
    private static let maxParseRetries = 3

    /// Base delay for parse retry backoff (seconds).
    private static let parseRetryBaseDelay: Double = 0.5

    // MARK: - Dependencies

    private let llmService: LLMService
    private let maxConcurrent: Int

    // MARK: - Initialization

    /// Initialize the parallel scoring service.
    ///
    /// - Parameters:
    ///   - llmService: The LLM service for making scoring requests.
    ///   - maxConcurrent: Maximum concurrent scoring requests.
    ///     Use `ConcurrencyDetector.detectConcurrency()` to determine
    ///     an appropriate value based on the provider.
    init(llmService: LLMService, maxConcurrent: Int) {
        self.llmService = llmService
        self.maxConcurrent = max(1, maxConcurrent)
    }

    // MARK: - Public API

    /// Score multiple documents in parallel.
    ///
    /// Uses a sliding window approach to maintain exactly `maxConcurrent`
    /// requests in flight at any time. Results are returned in completion
    /// order (not input order) for responsive progress updates.
    ///
    /// - Parameters:
    ///   - documents: Documents to score (as ScoringInput structs).
    ///   - claim: The medical claim to evaluate documents against.
    ///   - onProgress: Callback for progress updates.
    ///     - pmid: PMID of the just-completed document.
    ///     - completed: Number of documents completed so far.
    ///     - total: Total number of documents to score.
    /// - Returns: Array of scoring results in completion order.
    func scoreDocuments(
        _ documents: [ScoringInput],
        claim: String,
        onProgress: @escaping @Sendable (String, Int, Int) -> Void
    ) async -> [ScoringResult] {
        await scoreDocuments(documents, claim: claim, onProgress: onProgress, onResult: nil)
    }

    /// Score multiple documents in parallel with per-result callback.
    ///
    /// Uses a sliding window approach to maintain exactly `maxConcurrent`
    /// requests in flight at any time. Results are returned in completion
    /// order (not input order) for responsive progress updates.
    ///
    /// - Parameters:
    ///   - documents: Documents to score (as ScoringInput structs).
    ///   - claim: The medical claim to evaluate documents against.
    ///   - onProgress: Callback for progress updates.
    ///   - onResult: Optional callback for each result as it completes.
    ///     Called immediately when a document finishes scoring, enabling
    ///     incremental UI updates.
    /// - Returns: Array of scoring results in completion order.
    func scoreDocuments(
        _ documents: [ScoringInput],
        claim: String,
        onProgress: @escaping @Sendable (String, Int, Int) -> Void,
        onResult: (@Sendable (ScoringResult) async -> Void)?
    ) async -> [ScoringResult] {
        guard !documents.isEmpty else { return [] }

        var results: [ScoringResult] = []
        results.reserveCapacity(documents.count)
        var completed = 0
        let total = documents.count

        await withTaskGroup(of: ScoringResult.self) { group in
            var pending = documents[...]

            // Launch initial batch up to maxConcurrent
            for _ in 0..<min(maxConcurrent, documents.count) {
                if let doc = pending.popFirst() {
                    group.addTask {
                        await self.scoreDocument(doc, claim: claim)
                    }
                }
            }

            // Process results as they complete and refill the pool
            for await result in group {
                results.append(result)
                completed += 1
                onProgress(result.pmid, completed, total)

                // Call result handler for incremental updates
                await onResult?(result)

                // Add next document to maintain concurrency level
                if let doc = pending.popFirst() {
                    group.addTask {
                        await self.scoreDocument(doc, claim: claim)
                    }
                }
            }
        }

        return results
    }

    /// Aggregate total token usage from scoring results.
    ///
    /// - Parameter results: Array of scoring results.
    /// - Returns: Total input tokens, output tokens, and estimated cost.
    static func aggregateUsage(
        _ results: [ScoringResult]
    ) -> (inputTokens: Int, outputTokens: Int, estimatedCostUSD: Double) {
        var totalInput = 0
        var totalOutput = 0
        var totalCost: Double = 0

        for result in results {
            if let usage = result.usage {
                totalInput += usage.inputTokens
                totalOutput += usage.outputTokens
                totalCost += usage.estimatedCostUSD
            }
        }

        return (totalInput, totalOutput, totalCost)
    }

    // MARK: - Private Implementation

    /// Score a single document with retry logic for parse failures.
    ///
    /// - Parameters:
    ///   - input: The document scoring input.
    ///   - claim: The claim to evaluate against.
    /// - Returns: Scoring result with score/rationale or error.
    private func scoreDocument(
        _ input: ScoringInput,
        claim: String
    ) async -> ScoringResult {
        let prompt = buildScoringPrompt(input: input, claim: claim)
        let messages = [LLMService.userMessage(prompt)]

        var lastParseError = ""
        var accumulatedUsage: LLMUsage?

        // Retry loop for parse failures
        for attempt in 0..<Self.maxParseRetries {
            do {
                let (response, usage) = try await llmService.chat(
                    messages: messages,
                    temperature: 0.1,
                    maxTokens: 512,
                    jsonMode: true
                )

                // Accumulate usage across retries
                accumulatedUsage = accumulateUsage(accumulatedUsage, usage)

                // Parse the response
                let parsed = ResponseParser.parseScoreResponse(response)

                if !parsed.parseFailed, let score = parsed.score {
                    return .success(
                        pmid: input.pmid,
                        score: score,
                        rationale: parsed.explanation,
                        usage: accumulatedUsage
                    )
                }

                // Parse failed - prepare for retry
                lastParseError = parsed.explanation

                if attempt < Self.maxParseRetries - 1 {
                    let delay = Self.parseRetryBaseDelay * pow(2.0, Double(attempt))
                    let jitter = delay * Double.random(in: -0.25...0.25)
                    try await Task.sleep(for: .seconds(delay + jitter))
                }

            } catch {
                // Network/API error - return immediately (LLMService has its own retry logic)
                return .failure(
                    pmid: input.pmid,
                    error: error,
                    usage: accumulatedUsage
                )
            }
        }

        // All parse retries exhausted
        return .parseFailure(
            pmid: input.pmid,
            message: lastParseError,
            usage: accumulatedUsage
        )
    }

    /// Build the scoring prompt for a document.
    ///
    /// - Parameters:
    ///   - input: Document scoring input.
    ///   - claim: The claim to evaluate against.
    /// - Returns: Formatted prompt string.
    private func buildScoringPrompt(input: ScoringInput, claim: String) -> String {
        """
        Evaluate how relevant this document is to the following medical claim.

        Claim: \(claim)

        Document Title: \(input.title)
        Authors: \(input.authors)
        Year: \(input.year)
        Journal: \(input.journal)

        Abstract:
        \(input.abstract)

        IMPORTANT: Relevance means how useful the document is for ANSWERING the research question or EVALUATING the claim. Evidence that REFUTES or contradicts the claim is EQUALLY valuable as evidence that supports it. A study showing negative results is highly relevant if it directly addresses the claim.

        Score on a scale of 1-5:
        - 5: Directly addresses the claim with strong evidence (supporting OR refuting)
        - 4: Highly relevant, provides substantial information about the claim (positive or negative findings)
        - 3: Moderately relevant, contains useful related information
        - 2: Marginally relevant, tangentially related
        - 1: Not relevant to the claim

        Respond in JSON format only:
        {"score": <1-5>, "explanation": "<brief explanation>"}
        """
    }

    /// Accumulate usage from multiple LLM calls.
    ///
    /// - Parameters:
    ///   - existing: Previously accumulated usage (may be nil).
    ///   - new: New usage to add.
    /// - Returns: Combined usage.
    private func accumulateUsage(_ existing: LLMUsage?, _ new: LLMUsage) -> LLMUsage {
        guard let existing = existing else { return new }

        return LLMUsage(
            model: new.model,
            provider: new.provider,
            inputTokens: existing.inputTokens + new.inputTokens,
            outputTokens: existing.outputTokens + new.outputTokens
        )
    }
}
